shared/src/types/token.rs

```diff
@@ -297,6 +314,54 @@ pub fn is_non_owner_balance_key(key: &Key) -> Option<&Address> {
     }
 }
 
+/// Check if the given storage key is multitoken balance key for the given
+/// token. If it is, returns the sub prefix and the owner.
+pub fn is_multitoken_balance_key<'a>(
+    token_addr: &Address,
+    key: &'a Key,
+) -> Option<(Key, &'a Address)> {
+    match key.segments.first() {
+        Some(DbKeySeg::AddressSeg(addr)) if addr == token_addr => {
+            multitoken_balance_owner(key)
+        }
+        _ => None,
+    }
+}
+
+/// Check if the given storage key is multitoken balance key for unspecified
+/// token. If it is, returns the sub prefix and the owner.
+pub fn is_any_multitoken_balance_key(key: &Key) -> Option<(Key, &Address)> {
+    match key.segments.first() {
+        Some(DbKeySeg::AddressSeg(_)) => multitoken_balance_owner(key),
+        _ => None,
+    }
+}
+
+fn multitoken_balance_owner(key: &Key) -> Option<(Key, &Address)> {
+    let len = key.segments.len();
+    if len < 4 {
+        // the key of a multitoken should have 1 or more segments other than
+        // token, balance, owner
+        return None;
+    }
+    match key.get_at(len - 2) {
```

Comment:
you can use `?` and I think it will simplify the code a bit.

Reply:
Thanks! I tried to use a delimiter to simplify it.